### ISSUE TITLE
AI ignores patrolling mode while using Dimension Door or Town Portal spells

### ISSUE THREAD

**Issue by ihhub** [ihhub/fheroes2#7072]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Based on the feedback from players AI heroes completely ignore the fact that they are in patrol mode while using Dimension Door or Town Portal spells as AI heroes calculate distance based on move points, no on the number of tiles.

### Save file

I don't have one.

**Comment by ihhub** [comment]
Hi @Branikolog , please provide a save file for this.

**Comment by Branikolog** [comment]
Hi, @ihhub 
Here's a savefile, where AI patrol uses DD to jump and attack my hero. The hero doesn't jump beyond patrolling radius.
Here's save file. Just end the turn.
[Patrol teleport.zip](https://github.com/ihhub/fheroes2/files/11352955/Patrol.teleport.zip)
I couldn't reproduce TP issue. Trying to contact the user experienced TP issue now...

**Comment by oleg-derevenetz** [comment]
> The hero doesn't jump beyond patrolling radius.

So what's the problem then?

**Comment by Branikolog** [comment]
Hi, @oleg-derevenetz !
Problem with DD is that AI hero can jump behind some obstacle and then start wandering, searching his way back. I think allowing him to teleport is not safe, as he's leaving his position for a long time and could be abused (read further).
Between, I've managed to make patrol AI hero to use TP and he teleported to another castle and start making his business there.
![image](https://user-images.githubusercontent.com/55348946/235180984-3c29921b-92c8-4cf9-851c-23875fe7fe91.png)
I've blocked his way to his castle, so AI hero decided to TP. He occasionally chose more distant castle. His native village is located on the right side of this small island.
[AI TP trap.zip](https://github.com/ihhub/fheroes2/files/11354726/AI.TP.trap.zip)

**Comment by oleg-derevenetz** [comment]
@Branikolog

> Between, I've managed to make green patrol AI hero to use TP and he teleported to another castle and start making his business there.

But this hero in your save (Carlawn) doesn't have the `PATROL` mode:

```
******* Tile info *******
Tile index      : 1068, point: (24, 29)
UID             : 0
MP2 object type : 183 (Heroes)
ICN object type : 0 (UNKNOWN)
image index     : 255 (animated: 0)
layer type      : 0 - Object layer
region          : 4
ground          : Grass (isRoad: 0)
shadow          : false
passable from   : center, top, top-right, right, bottom-right, bottom, bottom-left, left, top-left
quantity 1      : 103
quantity 2      : 0
add. metadata   : 0
--- Extra information ---
name            : Carlawn (Sorceress)
color           : Green
experience      : 2500
level           : 3
magic points    : 34 / 40
position x, y   : 24, 29
move points     : 1000 / 1000
direction       : bottom-right
index sprite    : 27
in castle       : false
save object     : No object
flags           :
skills          : Expert Mysticism, Basic Wisdom, Basic Archery,
artifacts       : Magic Book, Spell Scroll,
spell book      : Bless, Shield, Cure, View Resources, Summon Boat, Disrupting Ray, Blind,
army dump       : color(Green), commander(Carlawn): 500 Peasants,
ai role         : Hunter
*************************
```

`flags` section doesn't show the `PATROL` mode and there is no information about the patrol zone.

**Comment by oleg-derevenetz** [comment]
> Could these magic numbers lie, in a case patrol hero went out of his zone? (which he actually did, while rushing treasures)

No, this mode is cleared only when hero has been lost one way or another (dismissed, lost in the battle or flee from battle).

> Problem with DD is that AI hero can jump behind some obstacle and then start wandering, searching his way back. I think allowing him to teleport is not safe, as he's leaving his position for a long time and could be abused (read further).

You might as well say that a hero on patrol might be blocked by wandering monsters during the Month of the Monster and using the dimension door will solve the problem in this case. In my opinion, hero can use any means of movement if they do not go beyond his patrol zone.

**Comment by Branikolog** [comment]
@oleg-derevenetz 
That's true. But I suppose the patrol hero should always keep his starting position. Otherwize, such maps, as Archibald 3 mission could be easily abused without even an army, capturing empty target castle, while patrol hero teleported somewhere and stuck there without mana or movement points. Can we really create an algorithm, that allows patrol hero teleport and then return back on the same turn? Counting his mana and movement points for that complex actions... I believe, for now we should better forbid such actions, untill AI could manage all that things in order to return back on the same turn he left his point.

**Comment by oleg-derevenetz** [comment]
@Branikolog I propose to figure out first whether there is a specific problem now, expressed in the fact that the hero on patrol goes beyond the patrol zone, or not :) Let's consider speculative theoretical problems separately.

**Comment by Branikolog** [comment]
Hi, @oleg-derevenetz 
I apologize, in this particular case I forgot to add patrol status for that hero. Now, after multiple tests I figured out, that patrol hero doesn't use Town Portal spell at all. Even he is blocked out of his patrol range.

But https://github.com/ihhub/fheroes2/issues/2315 is still valid, he's not designed to keep starting point.
Also, patrol hero can bypass some obstacles within his patrol range far beyond, which is not the case in the original game. Because of this Patrol hero can stuck (in a case of being blocked by creature from "month of", by other heroes or by boat stealing)
But these cases are out of scope of this issue, I suppose.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub is this issue still relevant? Investigation found no evidence that AI ignores the patrol zone using DD, TG or TP in patrol mode.

**Comment by ihhub** [comment]
Nope. I am closing it.
